File: content/LLD/Reference-Code/17. Decorater Pattern/good_example.py
from abc import ABC, abstractmethod


class Beverage(ABC):
    @abstractmethod
    def get_description(self) -> str:
        pass

    @abstractmethod
    def get_cost(self) -> int:
        pass


class Coffee(Beverage):
    def get_description(self):
        return "Plain coffee"

    def get_cost(self):
        return 20


class AddOnDecorator(Beverage):
    def __init__(self, coffee: "Coffee"):
        self._coffee = coffee

    def get_description(self):
        pass

    def get_cost(self):
        pass


class MilkDecorator(AddOnDecorator):
    def get_description(self):
        return self._coffee.get_description() + ", Milk"

    def get_cost(self):
        return self._coffee.get_cost() + 20


class WhipCreamDecorator(AddOnDecorator):
    def get_description(self):
        return self._coffee.get_description() + ", Whip Cream"

    def get_cost(self):
        return self._coffee.get_cost() + 50


class SugarDecorator(AddOnDecorator):
    def get_description(self):
        return self._coffee.get_description() + ", Sugar"

    def get_cost(self):
        return self._coffee.get_cost() + 5


coffee = Coffee()

coffee = MilkDecorator(coffee)
coffee = WhipCreamDecorator(coffee)
print(coffee.get_description())
print(coffee.get_cost())

File: content/LLD/Reference-Code/17. Decorater Pattern/GoodExample.java
public class GoodExample {
    interface Beverage {
        String getDescription();
        int getCost();
    }

    static class Coffee implements Beverage {
        @Override
        public String getDescription() {
            return "Plain coffee";
        }

        @Override
        public int getCost() {
            return 20;
        }
    }

    abstract static class AddOnDecorator implements Beverage {
        protected final Beverage coffee;

        public AddOnDecorator(Beverage coffee) {
            this.coffee = coffee;
        }
    }

    static class MilkDecorator extends AddOnDecorator {
        public MilkDecorator(Beverage coffee) {
            super(coffee);
        }

        @Override
        public String getDescription() {
            return this.coffee.getDescription() + ", Milk";
        }

        @Override
        public int getCost() {
            return this.coffee.getCost() + 20;
        }
    }

    static class WhipCreamDecorator extends AddOnDecorator {
        public WhipCreamDecorator(Beverage coffee) {
            super(coffee);
        }

        @Override
        public String getDescription() {
            return this.coffee.getDescription() + ", Whip Cream";
        }

        @Override
        public int getCost() {
            return this.coffee.getCost() + 50;
        }
    }

    static class SugarDecorator extends AddOnDecorator {
        public SugarDecorator(Beverage coffee) {
            super(coffee);
        }

        @Override
        public String getDescription() {
            return this.coffee.getDescription() + ", Sugar";
        }

        @Override
        public int getCost() {
            return this.coffee.getCost() + 5;
        }
    }

    public static void main(String[] args) {
        Beverage coffee = new Coffee();
        coffee = new MilkDecorator(coffee);
        coffee = new WhipCreamDecorator(coffee);
        System.out.println(coffee.getDescription());
        System.out.println(coffee.getCost());
    }
}
